x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java

```diff
@@ -363,10 +452,22 @@ public void getTrainedModel(final String modelId, final boolean includeDefinitio
                         String compressedString = docs.stream()
                             .map(TrainedModelDefinitionDoc::getCompressedString)
                             .collect(Collectors.joining());
-                        if (compressedString.length() != docs.get(0).getTotalDefinitionLength()) {
-                            listener.onFailure(ExceptionsHelper.serverError(
-                                Messages.getMessage(Messages.MODEL_DEFINITION_TRUNCATED, modelId)));
-                            return;
+                        // BWC for when we tracked the total definition length
+                        // TODO: remove in 9
+                        if (docs.get(0).getTotalDefinitionLength() != null) {
```

Comment:
Can the methods `getTrainedModelForInference` and `getTrainedModel` be refactored to avoid this duplication?

Reply:
definitely